packages/devtools_app/lib/src/shared/utils.dart

```diff
@@ -228,3 +230,51 @@ List<String> stripGoogle3(List<String> pathParts) {
 extension IsKeyType on KeyEvent {
   bool get isKeyDownOrRepeat => this is KeyDownEvent || this is KeyRepeatEvent;
 }
+
+/// A helper class for [Timer] functionality, where the callbacks are debounced.
+class DebounceTimer {
+  /// A periodic timer that ensures [callback] is only called at most once
+  /// per [duration].
+  ///
+  /// [callback] is triggered once immediately, and then every [duration] the
+  /// timer checks to see if the previous [callback] call has finished running.
+  /// If it has finished, then then next call to [callback] will begin.
+  DebounceTimer.periodic(
+    Duration duration,
+    Future<void> Function() callback,
+  ) {
+    unawaited(
+      callback().then((_) {
+        // If the timer was cancelled during the initial callback then don't start
+        // the periodic timer.
+        if (_isCancelled) return;
+
+        // Now a periodic timer is used to repeatedly trigger the callback after
+        // the duration.
+        _timer = Timer.periodic(duration, (timer) async {
+          // If the previous callback is still running, then don't trigger another
+          // callback. (debounce)
+          if (_isRunning) {
+            return;
+          }
+
```

Comment:
should we also check isCancelled here?

Reply:
The time would be cancelled before it is ever run, but this makes me realize I should probably start the timer at the same time as starting the callback so the duration gap is always the same! Thanks for asking this question!

Reply:
Your comment helped me realize that this could be written without _isCancelled all together :D